### ISSUE TITLE
Examples and usage not showing when using default command

### ISSUE THREAD

**Issue by qbolt** [yargs/yargs#1331]
Example:

```js
const Yargs = require('yargs')

Yargs.scriptName('my-app')
  .usage('Usage: $0 <command> [options]')
  .command(
    '$0',
    'Some description here',
    {},
    (argv) => console.log(argv)
  )
  .demandCommand()
  .recommendCommands()
  .strict()
  .example('$0', 'some example here')
  .showHelpOnFail(true)
  .help('?')
  .alias('?', ['help', 'h']).argv
```

If I make it a named command, the examples show. But by using the default command string specified in the documentation, it seems to omit the example usages.

Any help on this is greatly appreciated.

**Comment by bcoe** [comment]
@qbolt this sounds potentially like a bug! thanks for the report.

**Comment by jamesgeorge007** [comment]
@bcoe @qbolt I would like to take this up :clap: It would be great if you could point me to the right files to work upon as I'm new to the code-base :+1:

**Comment by bcoe** [comment]
@jamesgeorge007 would very much appreciate the help; I think this is probably going to be either in `usage.js` or in `command.js` (some interaction between these two files).

**Comment by WearyMonkey** [comment]
Any progress on this?

**Comment by bcoe** [comment]
@WearyMonkey haven't seen any updates, would still very much appreciate a patch.

**Comment by jamesgeorge007** [comment]
@bcoe Got busy with a couple of my personal projects. Also, didn't get much time to invest in familiarising with the code-base. It would be great if you could throw light on the specific file to work upon :clap:

**Comment by bcoe** [comment]
@jamesgeorge007 I'm not too sure, it's going to be somewhere in the relationship between, `command.js`, `yargs.js`, and `usage.js`.

I would probably start by looking in `yargs.js` and see what state everything is in as help is called.

**Comment by xmedeko** [comment]
I have s similar issue, probably. When using default command and `strict`, then when running with wrong params, yargs outputs just a help for the default command. But I would like to output full help like the `--help` does, i.e. to show the user all possible commands, too.

**Comment by mleguen** [comment]
@qbolt From what I understand of `.usage()`, it is not intended to be used when you defined a default command, as it itself results in performing a `.command()` call.

You should put the usage message as the 1st argument of the `.command()` call, for example:
```js
const Yargs = require('yargs')

Yargs.scriptName('my-app')
  .command(
    '$0 [options]',
    'Some description here',
    {},
    (argv) => console.log(argv)
  )
  .demandCommand()
  .recommendCommands()
  .strict()
  .showHelpOnFail(true)
  .help('?')
  .alias('?', ['help', 'h']).argv
```

Then it displays correctly as being the whole script usage:
```bash
my-app [options]

Some description here

Options:
  --version       Affiche le numéro de version                         [booléen]
  -?, --help, -h  Affiche de l'aide                                    [booléen]
```

**Comment by mleguen** [comment]
@qbolt Same thing for the examples, which should be defined in the default command builder:
```js
const Yargs = require('yargs')

Yargs.scriptName('my-app')
  .command(
    '$0 [options]',
    'Some description here',
    (y) => y.example('$0', 'some example here'),
    (argv) => console.log(argv)
  )
  .demandCommand()
  .recommendCommands()
  .strict()
  .showHelpOnFail(true)
  .help('?')
  .alias('?', ['help', 'h']).argv
```
Which results in:
```bash
my-app [options]

Some description here

Options:
  --version       Affiche le numéro de version                         [booléen]
  -?, --help, -h  Affiche de l'aide                                    [booléen]

Exemples:
  my-app  some example here
```

**Comment by mleguen** [comment]
@jamesgeorge007 If you are still willing to work on this issue, the work needed her IHMO is only modifying `docs/api.md` to add warnings in `.usage()` and `.example()` paragraphs about not using these functions if already defining a default command with `.command('$0...`

**Comment by mleguen** [comment]
@xmedeko Would you mind posting a code example and expected and actual outputs?

**Comment by xmedeko** [comment]
@mleguen Minimal example:
```js
const options = yargs
  .command(['cmd1', '*'], 'Command 1', { 'value': { demandOption: true } })
  .command('cmd2', 'Command 2', { 'value': { demandOption: true } })
  .help().strict(true).demandCommand(1)
  .parse();
```

Then run `node index.js foo` and output is (shortened):
```
Command 1

Options:
  --version  Show version number                       [boolean]
  --help     Show help                                          [boolean]
  --value                                                             [required]
```

But I would expect same similar to `node index.js --help`, i.e.
```
Command 1                                                                           
                                                                                    
Commands:                                                                           
  index.js cmd1  Command 1                                             [default]    
  index.js cmd2  Command 2                                                          
                                                                                    
Options:                                                                            
  --version  Show version number                       [boolean]    
  --help     Show help                                          [boolean]    
  --value                                                             [required]                                
```

So, that the user knows, the app has more commands.

**Comment by mleguen** [comment]
@xmedeko If I understand correctly what you mean, the behavior you want is to have other available commands displayed when the user asks for a command-specific help.

In that case, shouldn't your first (shortened) example be the output of `node index.js cmd1` (an existing command), not of `node index.js foo` (foo beeing an unknown command)?

We have no option in yargs at the moment to enable this behavior, and I am not sure we should add it, as it is not what most cli tools do (have a look at `git checkout --help` or `npm install --help` for example).

**Comment by xmedeko** [comment]
@mleguen Nope, when the user ask for a command specific help, then `yargs` behaves ok.

What I mean is, if you have a **default command**, and the user makes an **error in the args**, then `yargs` should not output usage for the default command (current behaviour). The problem is, the user may want to use another command, not the default, and the user does not know the app knows other commands.  `yargs` should output general help `node index.js --help` instead, it contains list of commands and args for the default command, too.

**Comment by mleguen** [comment]
@xmedeko OK, understood now. Mind opening a separate issue to track this feature request (something like "Default command help should display other available commands").
